File: 06.exs
# map_raw = "
# ....#.....
# .........#
# ..........
# ..#.......
# .......#..
# ..........
# .#..^.....
# ........#.
# #.........
# ......#...
# "
map_raw = File.read!("06.txt")

guard_dir = :up

# Parse into %{ {x, y} => char }
map_lines = String.split(map_raw, "\n", trim: true)

{map, guard_pos} =
  Enum.with_index(map_lines)
  |> Enum.reduce({%{}, nil}, fn {line, y}, {acc_map, arrow} ->
    Enum.with_index(String.graphemes(line))
    |> Enum.reduce({acc_map, arrow}, fn {char, x}, {map_acc, arrow_acc} ->
      new_map = Map.put(map_acc, {x, y}, char)

      new_arrow =
        if char == "^" do
          {x, y}
        else
          arrow_acc
        end

      {new_map, new_arrow}
    end)
  end)

defmodule Maze do
  def rotate(dir) do
    case dir do
      :up -> :right
      :right -> :down
      :down -> :left
      :left -> :up
    end
  end

  def step({map, {guard_x, guard_y}, guard_dir}) do
    next_pos_target =
      case guard_dir do
        :up -> {guard_x, guard_y - 1}
        :down -> {guard_x, guard_y + 1}
        :left -> {guard_x - 1, guard_y}
        :right -> {guard_x + 1, guard_y}
      end

    next_tile = Map.get(map, next_pos_target, nil)

    {next_pos_actual, next_dir} =
      case next_tile do
        nil -> {nil, guard_dir}
        "#" -> {{guard_x, guard_y}, rotate(guard_dir)}
        "." -> {next_pos_target, guard_dir}
      end

    map = Map.put(map, {guard_x, guard_y}, ".")
    map = Map.put(map, next_pos_actual, "^")
    # IO.inspect(next_pos_actual, label: "Guard Position")
    {map, next_pos_actual, next_dir}
  end

  def step_all({_, nil, _}, occupied, _) do
    {occupied, :escaped}
  end

  def step_all(state, occupied, prev_states) do
    {_, pos, dir} = state
    occupied = MapSet.put(occupied, pos)
    # we've been here before (incl. orientation) -> loop
    looped = MapSet.member?(prev_states, {pos, dir})

    if looped do
      {occupied, :looped}
    else
      prev_states = MapSet.put(prev_states, {pos, dir})
      next_state = step(state)
      step_all(next_state, occupied, prev_states)
    end
  end
end

# IO.inspect(guard_pos, label: "Guard Position")
{occupied, _} =
  {map, guard_pos, guard_dir}
  |> Maze.step_all(MapSet.new(), MapSet.new())

occupied
|> Enum.count()
|> IO.inspect(label: "A")

occupied
|> Enum.map(fn spot ->
  new_map = Map.put(map, spot, "#")
  {_, result} = Maze.step_all({new_map, guard_pos, guard_dir}, MapSet.new(), MapSet.new())
  result
end)
|> Enum.reject(&(&1 == :escaped))
|> Enum.count()
|> IO.inspect(label: "B")

File: 06.py
from pathlib import Path
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--full", action="store_true")
args = parser.parse_args()

if args.full:
    map_raw = Path("06.txt").read_text()
else:
    map_raw = """
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
"""

guard_dir = "up"

# Parse into {(x, y): char} and find guard position
map_lines = [line for line in map_raw.strip().split("\n")]

game_map = {}
guard_pos = None

for y, line in enumerate(map_lines):
    for x, char in enumerate(line):
        game_map[(x, y)] = char
        if char == "^":
            guard_pos = (x, y)

rotate_map = {
    "up": "right",
    "right": "down",
    "down": "left",
    "left": "up",
}


# Direction rotation
def rotate(dir):
    return rotate_map[dir]


# Step one move
def step(state):
    game_map, (guard_x, guard_y), guard_dir = state

    if guard_dir == "up":
        next_pos_target = (guard_x, guard_y - 1)
    elif guard_dir == "down":
        next_pos_target = (guard_x, guard_y + 1)
    elif guard_dir == "left":
        next_pos_target = (guard_x - 1, guard_y)
    elif guard_dir == "right":
        next_pos_target = (guard_x + 1, guard_y)
    else:
        raise ValueError(f"invalid direction {guard_dir}")

    next_tile = game_map.get(next_pos_target, None)

    if next_tile is None:
        next_pos_actual, next_dir = None, guard_dir
    elif next_tile == "#":
        next_pos_actual, next_dir = (guard_x, guard_y), rotate(guard_dir)
    elif next_tile == ".":
        next_pos_actual, next_dir = next_pos_target, guard_dir
    else:
        raise ValueError(f"invalid next_tile {next_tile}")

    # Mutate a copy of the map
    # new_map = dict(game_map)
    game_map[(guard_x, guard_y)] = "."
    game_map[next_pos_actual] = "^"

    return game_map, next_pos_actual, next_dir


# Recursive walking
def step_all(state: tuple, occupied: set, prev_states: set):
    while True:
        _, pos, dir = state

        if pos is None:
            return occupied, "escaped"

        occupied.add(pos)
        if (pos, dir) in prev_states:
            return occupied, "looped"

        prev_states.add((pos, dir))
        state = step(state)


# Part A
occupied, _ = step_all((game_map, guard_pos, guard_dir), set(), set())
print("A:", len(occupied))

# Part B
count = 0
for spot in occupied:
    new_map = dict(game_map)
    new_map[spot] = "#"
    _, result = step_all((new_map, guard_pos, guard_dir), set(), set())
    if result != "escaped":
        count += 1
print("B:", count)
